src/rdkafka_request.c

```diff
@@ -3328,6 +3328,284 @@ void rd_kafka_SaslAuthenticateRequest(rd_kafka_broker_t *rkb,
+        }
+err_parse:
+        return -1;
+}
+
+static int
+rd_kafkap_Produce_reply_tags_parse(rd_kafka_buf_t *rkbuf,
+                                   uint64_t tagtype,
+                                   uint64_t taglen,
+                                   rd_kafkap_Produce_reply_tags_t *tags) {
+        switch (tagtype) {
+        case 0: /* NodeEndpoints */
+                if (rd_kafka_buf_read_NodeEndpoints(rkbuf,
+                                                    &tags->NodeEndpoints) == -1)
+                        goto err_parse;
+                return 1;
+        default:
+                return 0;
+        }
+err_parse:
+        return -1;
+}
+
+/**
+ * @{
+ *
+ * @brief Leader discovery (KIP-951)
+ */
+
+void rd_kafkap_leader_discovery_tmpabuf_add_alloc_brokers(
+    rd_tmpabuf_t *tbuf,
+    rd_kafkap_NodeEndpoints_t *NodeEndpoints) {
+        int i;
+        size_t md_brokers_size =
+            NodeEndpoints->NodeEndpointCnt * sizeof(rd_kafka_metadata_broker_t);
+        size_t mdi_brokers_size = NodeEndpoints->NodeEndpointCnt *
+                                  sizeof(rd_kafka_metadata_broker_internal_t);
+        rd_tmpabuf_add_alloc_times(tbuf, md_brokers_size, 2);
+        rd_tmpabuf_add_alloc(tbuf, mdi_brokers_size);
+        for (i = 0; i < NodeEndpoints->NodeEndpointCnt; i++) {
+                size_t HostSize =
+                    RD_KAFKAP_STR_LEN(&NodeEndpoints->NodeEndpoints[i].Host) +
+                    1;
+                rd_tmpabuf_add_alloc(tbuf, HostSize);
+        }
+}
+
+void rd_kafkap_leader_discovery_tmpabuf_add_alloc_topics(rd_tmpabuf_t *tbuf,
+                                                         int topic_cnt) {
+        rd_tmpabuf_add_alloc(tbuf,
+                             sizeof(rd_kafka_metadata_topic_t) * topic_cnt);
+        rd_tmpabuf_add_alloc(tbuf, sizeof(rd_kafka_metadata_topic_internal_t) *
+                                       topic_cnt);
+}
+
+void rd_kafkap_leader_discovery_tmpabuf_add_alloc_topic(rd_tmpabuf_t *tbuf,
+                                                        char *topic_name,
+                                                        int32_t partition_cnt) {
+        if (topic_name) {
+                rd_tmpabuf_add_alloc(tbuf, strlen(topic_name) + 1);
+        }
+        rd_tmpabuf_add_alloc(tbuf, sizeof(rd_kafka_metadata_partition_t) *
+                                       partition_cnt);
+        rd_tmpabuf_add_alloc(tbuf,
+                             sizeof(rd_kafka_metadata_partition_internal_t) *
+                                 partition_cnt);
+}
+
+void rd_kafkap_leader_discovery_metadata_init(rd_kafka_metadata_internal_t *mdi,
+                                              int32_t broker_id) {
+        memset(mdi, 0, sizeof(*mdi));
+        mdi->metadata.orig_broker_id       = broker_id;
+        mdi->controller_id                 = -1;
+        mdi->cluster_authorized_operations = -1;
+}
+
+int rd_kafkap_leader_discovery_set_brokers(
```

Comment:
Change return type to void
```suggestion
void rd_kafkap_leader_discovery_set_brokers(
```

Reply:
Updated